python_coreml_stable_diffusion/torch2coreml.py

```diff
@@ -872,6 +940,169 @@ def cosine_distance(image_embeds, text_embeds):
+
+    if not hasattr(pipe, "text_encoder"):
+            raise RuntimeError(
+                "convert_text_encoder() deletes pipe.text_encoder to save RAM. "
+                "Please use convert_unet() before convert_text_encoder()")
+
+    for i, controlnet_model_version in enumerate(args.convert_controlnet):
+        controlnet_model_name = controlnet_model_version.replace("/", "_")
+        fname = f"ControlNet_{controlnet_model_name}.mlpackage"
+        out_path = os.path.join(args.o, fname)
+
+        if os.path.exists(out_path):
+            logger.info(
+                f"`controlnet_{controlnet_model_name}` already exists at {out_path}, skipping conversion."
+            )
+            continue
+
+        if i == 0:
+            batch_size = 2  # for classifier-free guidance
+            sample_shape = (
+                batch_size,                    # B
+                pipe.unet.config.in_channels,  # C
+                pipe.unet.config.sample_size,  # H
+                pipe.unet.config.sample_size,  # W
+            )
+
+            encoder_hidden_states_shape = (
+                batch_size,
+                pipe.text_encoder.config.hidden_size,
+                1,
+                pipe.text_encoder.config.max_position_embeddings,
+            )
+
+            controlnet_cond_shape = (
+                batch_size,                                           # B
+                3,                                                    # C
+                (args.latent_h or pipe.unet.config.sample_size) * 8,  # H
+                (args.latent_w or pipe.unet.config.sample_size) * 8,  # w
+            )
+
+            # Create the scheduled timesteps for downstream use
+            DEFAULT_NUM_INFERENCE_STEPS = 50
+            pipe.scheduler.set_timesteps(DEFAULT_NUM_INFERENCE_STEPS)
+
+            # Prepare inputs
+            sample_controlnet_inputs = OrderedDict([
+                ("sample", torch.rand(*sample_shape)),
+                ("timestep",
+                torch.tensor([pipe.scheduler.timesteps[0].item()] *
+                             (batch_size)).to(torch.float32)),
+                ("encoder_hidden_states", torch.rand(*encoder_hidden_states_shape)),
+                ("controlnet_cond", torch.rand(*controlnet_cond_shape)),
+            ])
+            sample_controlnet_inputs_spec = {
+                k: (v.shape, v.dtype)
+                for k, v in sample_controlnet_inputs.items()
+            }
+            logger.info(
+                f"Sample ControlNet inputs spec: {sample_controlnet_inputs_spec}")
+
+            baseline_sample_controlnet_inputs = deepcopy(sample_controlnet_inputs)
+            baseline_sample_controlnet_inputs[
+                "encoder_hidden_states"] = baseline_sample_controlnet_inputs[
+                    "encoder_hidden_states"].squeeze(2).transpose(1, 2)
+
+        # Import controlnet model and initialize reference controlnet
+        original_controlnet = ControlNetModel.from_pretrained(
```

Comment:
I reviewed https://huggingface.co/lllyasviel/sd-controlnet-mlsd/blob/main/config.json to see if there was an indication of which `StableDiffusionModel` models a particular `ControlNetModel` model is compatible with but didn't find anything. Do you have a recommended way perform such a compatibility check?

Reply:
Hmm, comparing the shapes of ControlNet outputs with those of UNet additional inputs to ensure they are the same may work as a part of the compatibility check, but it does not completely assure compatibility.
I can check them while converting models in torch2coreml.py and warn if they are not compatible.

Reply:
I meant more of a `config.json` level compatibility hint but it seems that this information does not currently exist. @pcuenca what do you think? This is a nice-to-have (not a merge blocker in my opinion)

Reply:
Thanks for the ping! I think it's a good idea to save the base model in the `config.json` of the trained ControlNet models, I'll create an issue in `diffusers` for discussion.

Regarding the initial collection of ControlNet models, as far as I know they were all trained using [Stable Diffusion v1.5](https://huggingface.co/runwayml/stable-diffusion-v1-5), so that could be a reasonable fallback default.

Reply:
Thanks @pcuenca ! We can update this repo once that issue is resolved on the diffusers side. @ryu38 For the short term (merging this PR before diffusers issue is resolved), could you please add a `logger.warning` message mentioning the version compatibility risk and maybe call out v1.5 as the default for the ControlNet original paper models?

Reply:
I added the warning message as below in the new commit.
https://github.com/apple/ml-stable-diffusion/blob/f535554024706602f66c65a5e8b4b6205899afc3/python_coreml_stable_diffusion/torch2coreml.py#L956-L959

Reply:
Great!